src/markov/mc_tools.jl

```diff
@@ -57,172 +59,217 @@ end
 
 function Base.show{T,TM}(io::IO, mc::MarkovChain{T,TM})
     println(io, "Discrete Markov Chain")
     println(io, "stochastic matrix of type $TM:")
-    println(io, mc.p)
+    print(io, mc.p)
 end
 
 """
-solve x(P-I)=0 using an algorithm presented by Grassmann-Taksar-Heyman (GTH)
+This routine computes the stationary distribution of an irreducible Markov
+transition matrix (stochastic matrix) or transition rate matrix (generator
+matrix) `A`.
+
+More generally, given a Metzler matrix (square matrix whose off-diagonal
+entries are all nonnegative) `A`, this routine solves for a nonzero solution
+`x` to `x (A - D) = 0`, where `D` is the diagonal matrix for which the rows of
+`A - D` sum to zero (i.e., `D_{ii} = \sum_j A_{ij}` for all `i`). One (and only
+one, up to normalization) nonzero solution exists corresponding to each
+reccurent class of `A`, and in particular, if `A` is irreducible, there is a
+unique solution; when there are more than one solution, the routine returns the
+solution that contains in its support the first index `i` such that no path
+connects `i` to any index larger than `i`. The solution is normalized so that
+its 1-norm equals one. This routine implements the Grassmann-Taksar-Heyman
+(GTH) algorithm (Grassmann, Taksar, and Heyman 1985), a numerically stable
+variant of Gaussian elimination, where only the off-diagonal entries of `A` are
+used as the input data. For a nice exposition of the algorithm, see Stewart
+(2009), Chapter 10.
 
 ##### Arguments
 
-- `p::Matrix` : valid stochastic matrix
+- `A::Matrix{T}` : Stochastic matrix or generator matrix. Must be of shape n x
+  n.
 
 ##### Returns
 
-- `x::Matrix`: A matrix whose columns contain stationary vectors of `p`
+- `x::Vector{T}` : Stationary distribution of `A`.
 
 ##### References
 
-The following references were consulted for the GTH algorithm
-
 - W. K. Grassmann, M. I. Taksar and D. P. Heyman, "Regenerative Analysis and
 Steady State Distributions for Markov Chains, " Operations Research (1985),
 1107-1116.
 - W. J. Stewart, Probability, Markov Chains, Queues, and Simulation, Princeton
 University Press, 2009.
 
 """
-gth_solve{T<:Integer}(a::AbstractMatrix{T}) =
-    gth_solve(convert(AbstractArray{Rational, 2}, a))
-
-# solve x(P-I)=0 by the GTH method
-function gth_solve{T<:Real}(original::AbstractMatrix{T})
-    # NOTE: the full here will convert a sparse matrix to dense before the gth
-    #       algorithm begins. Given the nature of the algorithm this will
-    #       almost certainly be more efficient than operating on the sparse
-    #       matrix itself.
-    a = copy(full(original))
-    n = size(a, 1)
+gth_solve{T<:Real}(A::Matrix{T}) = gth_solve!(copy(A))
+
+# convert already makes a copy, hence gth_solve!
+gth_solve{T<:Integer}(A::Matrix{T}) =
+    gth_solve!(convert(Matrix{Rational{T}}, A))
+
+"""
+Same as `gth_solve`, but overwrite the input `A`, instead of creating a copy.
+
+"""
+function gth_solve!{T<:Real}(A::Matrix{T})
```

Comment:
I was thinking just keeping in general. 

I'm not sure what the benefits would be, but it seems easy enough to keep that option open for the future